From 0b9b729a9001a4e59f39e68b73ebc5c26d29ea34 Mon Sep 17 00:00:00 2001
From: ywan171 <yi.a.wang@intel.com>
Date: Thu, 7 Feb 2013 16:03:51 +0800
Subject: libmediaplayerservice: shutdown video decoder when it has error and wait discontinuity

BZ: 81083

when video decoder's loop stops after video omx client reports error
if it needs discontinuity flag then, it can not get it. Shut down video decoder
in this condition to avoid reset ANR.

Category: aosp improvement
Domain: Video.Media-streaming
Origin: internal
Upstream-Candidate: yes

Change-Id: I83b6ba4d52e09a5a9dca487ebb0d064e17dd461e
Orig-Change-Id: Id868e3140b0f9bb476c435f57009103bfe59c982
Signed-off-by: ywan171 <yi.a.wang@intel.com>
Signed-off-by: chuzhoux <chux.zhou@intel.com>
---
 media/libmediaplayerservice/nuplayer/NuPlayer.cpp |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/media/libmediaplayerservice/nuplayer/NuPlayer.cpp b/media/libmediaplayerservice/nuplayer/NuPlayer.cpp
index 44fb626..c277fee 100644
--- a/media/libmediaplayerservice/nuplayer/NuPlayer.cpp
+++ b/media/libmediaplayerservice/nuplayer/NuPlayer.cpp
@@ -780,6 +780,12 @@ void NuPlayer::onMessageReceived(const sp<AMessage> &msg) {
                     ALOGE("%s track encountered an error (%d)",
                          audio ? "audio" : "video", finalResult);
 
+                    if (!audio && (mVideoDecoder != NULL)
+                            && (mFlushingVideo == AWAITING_DISCONTINUITY)) {
+                        // if video decoder is in AWAITING_DISCONTINUITY state
+                        // shut down it to avoid reset blocking
+                        flushDecoder(false,true);
+                    }
                     notifyListener(
                             MEDIA_ERROR, MEDIA_ERROR_UNKNOWN, finalResult);
                 }
-- 
1.7.4.1

